### PR TITLE
Optimizations for Resnet50

### PR DESCRIPTION
**Pull request by gautamr-groq**
Redundant cast optimization moved to Groq codebase because it can lead to subtle issues. Only implement distribute for mult optimizations

### CODE REVIEW

File: src/Transform/ONNX/ConstProp.td [hunk 1]
```diff
@@ -185,6 +187,25 @@ def SqrtofConst :  Pat<
...
 //===----------------------------------------------------------------------===//
 
+def ReduceZeroAdd : Pat<
+  // From add(x, 0).
+  (ONNXAddOp $x,(ONNXConstantOp:$c1 $s1, $v1)),
+  // To x
+  (replaceWithValue $x),
+  [(isDenseZeroTensor $v1)]>;
+
+//something off about the shapes
```
Comment (gautamr-groq) [inline src/Transform/ONNX/ConstProp.td]: accidentally left this comment in here, can remove now

File: src/Transform/ONNX/Combine.td [hunk 2]
```diff
@@ -75,6 +75,12 @@ def CastEliminationPattern : Pat<
 	(replaceWithValue $arg),
   [(HasSameElementType $arg, $type)]>;
 
+// (onnx.Cast (onnx.Cast(%X, $type2), $type1)) = (onnx.Cast(%X, $type1))
+def CastPropagationPattern : Pat<
```
Comment (kkang-groq) [inline src/Transform/ONNX/Combine.td]: As we talked about, this affects result if we optimize away lower precision type.
For example, if we had `cast<fp32>(cast<int8>(x))`, this would not produce same result as `cast<fp32>(x)`.
This might be okay if we were doing static quantization, since this might improve accuracy.
However, if the model was trained using quantization aware training, the training model would have taken clip to [-128, 127] of cast<int8>(x) into account in its parameters, and optimizing away that operation may produce lower accuracy results.

File: src/Transform/ONNX/ConstProp.td [hunk 3]
```diff
@@ -185,6 +187,25 @@ def SqrtofConst :  Pat<
 // Exactly the same pattern as for the elementwise ADD operations.
 //===----------------------------------------------------------------------===//
 
+def ReduceZeroAdd : Pat<
```
Comment (kkang-groq) [inline src/Transform/ONNX/ConstProp.td]: Add test case where x is broadcasted to dense zero tensor's shape, followed by MatMul.
For example, something like:
```
%2 = onnx.constant()  {value = dense<0> : <320xf32>} : () -> tensor<320xf32>
%0 = onnx.add(%1, %2) : (tensor<1xf32>, tensor<320xf32>) -> tensor<320xf32>
%3 = TMatMul(%0, %4) : (tensor<320xf32>, tensor<320xf32>) -> tensor<1xf32>
```

Comment (kkang-groq) [reply src/Transform/ONNX/ConstProp.td]: If above case fails, we might want to only do ReduceZeroAdd if x doesn't broadcast to zero tensor's shape.

File: src/Transform/ONNX/Combine.td (hunk 2 above)
Comment (gautamr-groq) [reply src/Transform/ONNX/Combine.td]: Thanks for pointing this out. I'm moving this into our codebase and creating a follow up item to investigate if this makes our true accuracy better or worse.

File: src/Transform/ONNX/ConstProp.td (hunk 3 above)
Comment (gautamr-groq) [reply src/Transform/ONNX/ConstProp.td]: You are right that test fails, I'm just going to get rid of the add by 0 in quantize/dequantize lowering.

File: src/Transform/ONNX/ConstProp.td [hunk 4]
```diff
@@ -185,6 +187,25 @@ def SqrtofConst :  Pat<
...
+  // From add(x, 0).
+  (ONNXAddOp $x,(ONNXConstantOp:$c1 $s1, $v1)),
+  // To x
+  (replaceWithValue $x),
+  [(isDenseZeroTensor $v1)]>;
+
+//something off about the shapes
+def MulAddDistribitive : Pat<
```
Comment (kkang-groq) [inline src/Transform/ONNX/ConstProp.td:188]: When do you know the transformed version is better than the original?
Assuming ONNXMulOp of the two constants is done at compile time, both are 2 ops. One does add followed by mul, the other does vice versa.
Does ONNX MLIR figure out which of the two cases can do more constant propagation?

Comment (gautamr-groq) [reply src/Transform/ONNX/ConstProp.td:188]: The Mul of the 2 constants does happen at compile time. In general this reduces a chain of (C(B(Ax+a)+ b)+c) ... into a operation like Ax+b. I don't think that chain is currently reduced by any of the other other ONNX Const Propagations. The current ONNX const propagations optimize away associative additions and multiplications and this change does not affect that. If anything it can only create more associative additions (like a(x+b) + c(y+d) = ax + cy + (ab+cd))

File: src/Transform/ONNX/ConstProp.td (hunk 3 above)
Comment (gautamr-groq) [reply src/Transform/ONNX/ConstProp.td]: Also added the test into the end_to_end tests in this mr : https://git.groq.io/code/Groq/-/merge_requests/9322

File: src/Transform/ONNX/ConstProp.td (hunk 4 above)
Comment (kkang-groq) [reply src/Transform/ONNX/ConstProp.td:188]: Ah I see. Makes sense.

**Comment by kkang-groq** [comment]
LGTM! 👍
